### PR TITLE
Support for assisted project updates

### PR DESCRIPTION
**Pull request by aloubyansky**
Introduces two dev tools commands:
* info - logging the current project Quarkus-related info such as imported platform BOMs, and direct (and top-level transitive) extension dependencies, possibly indicating version misalignment;
* update - queries enabled registries, calculates recommended updates and logs the recommendations.

For example
````
[aloubyansky@localhost code-with-quarkus]$ mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:info
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------< org.acme:code-with-quarkus >---------------------
[INFO] Building code-with-quarkus 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:info (default-cli) @ code-with-quarkus ---
[INFO] Quarkus platform BOMs:
[INFO]   io.quarkus.platform:quarkus-bom:pom:2.5.2.Final
[INFO]   io.quarkus.platform:quarkus-camel-bom:pom:2.4.1.Final | unnecessary
[INFO] 
[INFO] Extensions from registry.quarkus.io:
[INFO]   io.quarkiverse.prettytime:quarkus-prettytime:0.2.1
[INFO] 
[INFO] Extensions from io.quarkus.platform:quarkus-kogito-bom:
[INFO]   org.kie.kogito:kogito-quarkus-rules:2.0.0-SNAPSHOT | misaligned
[INFO] 
[INFO] Extensions from io.quarkus.platform:quarkus-bom:
[INFO]   io.quarkus:quarkus-hibernate-orm:2.6.2.Final | misaligned
[INFO]   io.quarkus:quarkus-openshift
[INFO]   io.quarkus:quarkus-arc
[INFO] 
[WARNING] Non-recommended Quarkus platform BOM and/or extension versions were found. For more details, please, execute 'mvn quarkus:update -Drectify'
````
````
[aloubyansky@localhost code-with-quarkus]$ mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:update -Drectify
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------< org.acme:code-with-quarkus >---------------------
[INFO] Building code-with-quarkus 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:update (default-cli) @ code-with-quarkus ---
[INFO] Looking for the newly published extensions in registry.quarkus.redhat.com
[INFO] Quarkus platform BOMs:
[INFO]         io.quarkus.platform:quarkus-bom:pom:2.5.2.Final
[INFO] Remove: io.quarkus.platform:quarkus-camel-bom:pom:2.4.1.Final
[INFO] Add:    io.quarkus.platform:quarkus-kogito-bom:pom:2.5.2.Final
[INFO] 
[INFO] Extensions from registry.quarkus.io:
[INFO]   io.quarkiverse.prettytime:quarkus-prettytime:0.2.1
[INFO] 
[INFO] Extensions from io.quarkus.platform:quarkus-kogito-bom:
[INFO] Update: org.kie.kogito:kogito-quarkus-rules:2.0.0-SNAPSHOT -> org.kie.kogito:kogito-quarkus-rules
[INFO] 
[INFO] Extensions from io.quarkus.platform:quarkus-bom:
[INFO] Update: io.quarkus:quarkus-hibernate-orm:2.6.2.Final -> io.quarkus:quarkus-hibernate-orm
[INFO]         io.quarkus:quarkus-openshift
[INFO]         io.quarkus:quarkus-arc
````
With the `code.quarkus.redhat.com` registry enabled:
````
[aloubyansky@localhost code-with-quarkus]$ mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:update
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------< org.acme:code-with-quarkus >---------------------
[INFO] Building code-with-quarkus 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:update (default-cli) @ code-with-quarkus ---
[INFO] Looking for the newly published extensions in registry.quarkus.redhat.com
[INFO] 
[INFO] Recommended Quarkus platform BOM updates:
[INFO] Add:    com.redhat.quarkus.platform:quarkus-bom:pom:2.2.3.SP1-redhat-00002
[INFO] Add:    io.quarkus.platform:quarkus-kogito-bom:pom:2.2.3.Final
[INFO] Remove: io.quarkus.platform:quarkus-bom:pom:2.5.2.Final
[INFO] Remove: io.quarkus.platform:quarkus-camel-bom:pom:2.4.1.Final
[INFO] 
[INFO] Extensions from com.redhat.quarkus.platform:quarkus-bom:pom:2.2.3.SP1-redhat-00002:
[INFO] Update: io.quarkus:quarkus-hibernate-orm:2.6.2.Final -> io.quarkus:quarkus-hibernate-orm
[INFO] 
[INFO] Extensions from io.quarkus.platform:quarkus-kogito-bom:pom:2.2.3.Final:
[INFO] Update: org.kie.kogito:kogito-quarkus-rules:2.0.0-SNAPSHOT -> org.kie.kogito:kogito-quarkus-rules
[INFO] 
[INFO] Extensions from registry.quarkus.io:
[INFO] Update: io.quarkiverse.prettytime:quarkus-prettytime:0.2.1 -> 0.1.2
````
Equivalent `quarkusInfo` and `quarkusUpdate` Gradle commands are included.

### CODE REVIEW

**Comment by aloubyansky** [comment]
I will still add a few things. But generally, this is the first version of it.

**Comment by aloubyansky** [comment]
Here is a per-module option for a multi-module project
````
[aloubyansky@localhost runner]$ mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:info -DperModule
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< org.acme:acme-runner >------------------------
[INFO] Building acme-runner 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:info (default-cli) @ acme-runner ---
[INFO] Looking for the newly published extensions in registry.quarkus.io
[INFO] Quarkus platform BOMs:
[INFO]   io.quarkus:quarkus-bom:pom:999-SNAPSHOT
[INFO] 
[INFO] Main application module org.acme:acme-runner:
[INFO]   Build file: pom.xml
[INFO]   Extensions from io.quarkus:quarkus-bom:
[INFO]     io.quarkus:quarkus-resteasy:2.6.2.Final | misaligned
[INFO] 
[INFO] Module org.acme:acme-beans:
[INFO]   Build file: ../beans/pom.xml
[INFO]   Extensions from io.quarkus:quarkus-bom:
[INFO]     io.quarkus:quarkus-arc
[INFO] 
[WARNING] Non-recommended Quarkus platform BOM and/or extension versions were found. For more details, please, execute 'mvn quarkus:update -Drectify'
````

**Review by gastaldi — APPROVED** [review]
LGTM

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/UpdateCommandHandler.java [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+                    sb.append(String.format(UpdateCommandHandler.PLATFORM_RECTIFY_FORMAT,
+                            UpdateCommandHandler.UPDATE, c.toCompactCoords()));
+                    sb.append(" -> ");
+                    sb.append(c.getGroupId()).append(':')
+                            .append(c.getArtifactId());
+                    if (!c.getClassifier().isEmpty()) {
+                        sb.append(':').append(c.getClassifier());
+                    }
+                    log.info(sb.toString());
+                }
```
Comment (ebullient) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/UpdateCommandHandler.java:178]: I think it was this one. If the recommendation is to remove the version (shown in the original snippet for the item), the display: GAV -> GA .. doesn't explain to me what I should do (which is remove the version specifier).

**Review by ebullient — APPROVED** [review]

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/UpdateCommandHandler.java (hunk 1 above)
Comment (aloubyansky) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/UpdateCommandHandler.java:173]: I've changed to ` -> remove version (managed)` for now as discussed.

**Review by gsmet — APPROVED** [review]
Let's get this in and adjust during the CR phase.

**Review by glefloch — APPROVED** [review]
LGTM

**Comment by aloubyansky** [comment]
Added some javadoc and formatting adjustment.

**Review by ebullient — APPROVED** [review]

**Comment by famod** [comment]
Looks related to #7272?

**Comment by aloubyansky** [comment]
Yes, most of the points are going to be covered in 2.7. This is the first iteration. Improvements and refactoring will follow.

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusInfo.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+public class QuarkusInfo extends QuarkusPlatformTask {
+
+    private boolean perModule = false;
+
+    @Input
+    public boolean getPerModule() {
+        return perModule;
+    }
+
+    @Option(description = "Log project's state per module.", option = "perModule")
```
Comment (maxandersen) [inline devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusInfo.java:26]: didn't you want to hide/remove this option for this round?

agree on its purpose, just not sold on its name :)

File: devtools/maven/src/main/java/io/quarkus/maven/QuarkusProjectStateMojoBase.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    boolean perModule;
+
+    @Override
+    public void doExecute(QuarkusProject quarkusProject, MessageWriter log) throws MojoExecutionException {
+        if (project.getFile() == null) {
+            throw new MojoExecutionException("This goal requires a project");
+        }
+
+        if (!QuarkusProjectHelper.isRegistryClientEnabled()) {
+            throw new MojoExecutionException("This goal requires a Quarkus extension registry client to be enabled");
```
Comment (maxandersen) [inline devtools/maven/src/main/java/io/quarkus/maven/QuarkusProjectStateMojoBase.java:50]: if this triggers what is a user to do about it ?

**Comment by glefloch** [comment]
Next step would be to add this in the CLI?

**Review by maxandersen — CHANGES_REQUESTED** [review]
Looks good as a start - a few observations:

1) As this is provisional/experimental I feel we are best off printing a info/warning at the top:
"This update|info command is experimental. flags, options and output is bound to change." or similar.

 2) I get "Looking for the newly published extensions in registry.quarkus.io" printed twice.
```
[INFO] Looking for the newly published extensions in registry.quarkus.io
[INFO] Looking for the newly published extensions in registry.quarkus.io
```
Guessing its one for each registry and the name is not right ? one should be registry.quarkus.redhat.com ?

3) The spaces/newlines makes a big difference and the add/remove works well

4) Main missing thing in this I feel are examples/testcases to try it out on ...for now just doing random experiments but I'm sure I'm missing some. Not something to fix for today but definitely should add that sooner rather than later.

5)  Instead of "Recommended Quarkus platform BOM updates" should we write "To follow latest recommended Quarkus platform:" ?

6) in general the language used is a bit "terse" which I don't have a good answer to as my own versions would just be different kind of "terse" - something when we have some more specific testcases should sit down and go through and see if we can find better wording. hence #1 is important too :)

7) cli support we can live without but should definitely be in before final; even if just passing directly through.

I'll play around a bit more but for now, I'm +1 on merging it when we get #1 addressed.

**Comment by maxandersen** [comment]
here is a weird one:

```
qssnap create -P io.quarkus:quarkus-bom:999-SNAPSHOT  # create using snaphost quarkus core
qssnap ext add cxf # not sure this matters
mvn quarkus:update # tells me to use recommended platform as expected
```

then I change it to use the recommended platform:

```
<quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
    <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
    <quarkus.platform.version>2.6.2.Final</quarkus.platform.version>
```

then run `mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:update` since `quarkus:update` does not exist and I get:

```
[INFO] Recommended Quarkus platform BOM updates:
[INFO] Update: io.quarkus.platform:quarkus-bom:pom:2.6.2.Final -> 2.6.2.Final
```

That does not sound right ?

info gives me:

```
[INFO] Quarkus platform BOMs:
[INFO]   io.quarkus.platform:quarkus-bom:pom:2.6.2.Final
[INFO]
[INFO] Extensions from registry.quarkus.io:
[INFO]   io.quarkiverse.cxf:quarkus-cxf:0.12.0
[INFO]
[INFO] Extensions from io.quarkus.platform:quarkus-bom:
[INFO]   io.quarkus:quarkus-arc
[INFO]   io.quarkus:quarkus-resteasy
```

**Review by maxandersen — APPROVED** [review]
discucsing with @aloubyansky - okey with merging this; we can add the warning after CR1.

**Comment by aloubyansky** [comment]
> here is a weird one:
> 
> ```
> qssnap create -P io.quarkus:quarkus-bom:999-SNAPSHOT  # create using snaphost quarkus core
> qssnap ext add cxf # not sure this matters
> mvn quarkus:update # tells me to use recommended platform as expected
> ```
> 
> then I change it to use the recommended platform:
> 
> ```
> <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
>     <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
>     <quarkus.platform.version>2.6.2.Final</quarkus.platform.version>
> ```
> 
> then run `mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:update` since `quarkus:update` does not exist and I get:
> 
> ```
> [INFO] Recommended Quarkus platform BOM updates:
> [INFO] Update: io.quarkus.platform:quarkus-bom:pom:2.6.2.Final -> 2.6.2.Final
> ```
> 
> That does not sound right ?
> 
> info gives me:
> 
> ```
> [INFO] Quarkus platform BOMs:
> [INFO]   io.quarkus.platform:quarkus-bom:pom:2.6.2.Final
> [INFO]
> [INFO] Extensions from registry.quarkus.io:
> [INFO]   io.quarkiverse.cxf:quarkus-cxf:0.12.0
> [INFO]
> [INFO] Extensions from io.quarkus.platform:quarkus-bom:
> [INFO]   io.quarkus:quarkus-arc
> [INFO]   io.quarkus:quarkus-resteasy
> ```

Right, thanks.

**Comment by aloubyansky** [comment]
I opened https://github.com/quarkusio/quarkus/pull/22999 to address 1, 2 and the update to the current version issue. I'll be adding a CLI command and tests too.

**Comment by gsmet** [comment]
@aloubyansky could you write something I could include in the announcement to present the feature? A comment here is fine.

**Comment by rsvoboda** [comment]
@aloubyansky is there any text for assisted updates? I didn't find anything in the blog post announce.

**Comment by aloubyansky** [comment]
There isn't any. I was working on a proper testsuite and thought I would add it with that. I did not forget about it.

**Comment by aloubyansky** [comment]
This is what we have so far:
https://quarkus.io/guides/maven-tooling#project-info
https://quarkus.io/guides/maven-tooling#project-update

**Pull request merged**